File: ts/src/prober.ts
/*
 *  Description: The measurement loop: cheap and constant for latency, rare and careful for
 *               throughput.
 *
 *               Two rules shape everything here.
 *
 *               A probe must not distort what it measures. Latency probes are tiny and spaced out;
 *               the throughput probe downloads real bytes, so it only runs when the application has
 *               been quiet for a while — measuring bandwidth during a burst of user traffic would
 *               compete with that traffic and report a number caused by the measurement itself.
 *
 *               A failing line must not be probed at the same rate as a healthy one. A line that is
 *               down gets exponential backoff, so a dead route costs a request every few minutes
 *               rather than every few seconds.
 *
 *  Author(s):
 *      agent4
 */

import type { HealthTable } from "./health.js";
import type { Line } from "./registry.js";

export interface ProberOptions {
  /** Path of the unauthenticated liveness endpoint. */
  readonly probePath?: string;
  /** Gap between latency probes of a healthy line. */
  readonly intervalMs?: number;
  /** A probe that has not answered within this is a failure. */
  readonly timeoutMs?: number;
  /** Backoff cap for a line that keeps failing. */
  readonly maxBackoffMs?: number;

  /** Path of a download used to measure throughput. Absent disables throughput probing entirely. */
  readonly bandwidthPath?: string;
  /** Minimum gap between throughput measurements of the same line. */
  readonly bandwidthIntervalMs?: number;
  /** How long the application must have been quiet before a throughput probe may run. */
  readonly idleBeforeBandwidthMs?: number;

  /** Injected in tests. */
  readonly now?: () => number;
}

const DEFAULTS = {
  probePath: "/mt/probe",
  intervalMs: 15_000,
  timeoutMs: 5_000,
  maxBackoffMs: 300_000,
  // Conservative on purpose, and provisional: the design doc leaves these to be settled against
  // real lines. Ten minutes between measurements of a number that changes slowly, after half a
  // minute of quiet, downloading a quarter of a megabyte — small enough not to matter on a metered
  // connection, large enough to outlast connection setup and actually measure the pipe.
  bandwidthIntervalMs: 600_000,
  idleBeforeBandwidthMs: 30_000,
} as const;

/**
 * Drives the health table.
 *
 * Deliberately not started by the constructor: probing is a background activity with a real cost,
 * and a library that begins making network requests the moment it is instantiated is a library
 * that surprises people. The caller says when.
 */
export class Prober {
  private readonly options: Required<Omit<ProberOptions, "bandwidthPath" | "now">> & {
    bandwidthPath: string | undefined;
    now: () => number;
  };
  private timer: ReturnType<typeof setTimeout> | null = null;
  private running = false;
  /** Next allowed probe time per line, which is how backoff is expressed. */
  private readonly nextProbeAt = new Map<string, number>();
  private readonly nextBandwidthAt = new Map<string, number>();
  /** When the application last issued a real request, so quiet can be detected. */
  private lastTrafficAt = 0;

  constructor(
    private readonly lines: () => readonly Line[],
    private readonly health: HealthTable,
    private readonly fetchImpl: typeof globalThis.fetch,
    private readonly resolve: (path: string, line: Line) => string,
    options: ProberOptions = {},
  ) {
    this.options = {
      probePath: options.probePath ?? DEFAULTS.probePath,
      intervalMs: options.intervalMs ?? DEFAULTS.intervalMs,
      timeoutMs: options.timeoutMs ?? DEFAULTS.timeoutMs,
      maxBackoffMs: options.maxBackoffMs ?? DEFAULTS.maxBackoffMs,
      bandwidthIntervalMs: options.bandwidthIntervalMs ?? DEFAULTS.bandwidthIntervalMs,
      idleBeforeBandwidthMs: options.idleBeforeBandwidthMs ?? DEFAULTS.idleBeforeBandwidthMs,
      bandwidthPath: options.bandwidthPath,
      now: options.now ?? (() => Date.now()),
    };
  }

  /** Called by the line manager on every real request, so the prober can tell quiet from busy. */
  noteTraffic(): void {
    this.lastTrafficAt = this.options.now();
  }

  start(): void {
    if (this.running) return;
    this.running = true;
    void this.tick();
  }

  stop(): void {
    this.running = false;
    if (this.timer !== null) clearTimeout(this.timer);
    this.timer = null;
  }

  /** Probe everything now and wait for it. The panel's refresh button, and every test's lever. */
  async probeAll(): Promise<void> {
    await Promise.all(this.lines().map((line) => this.probe(line)));
    this.health.reconcileDegraded();
  }

  private async tick(): Promise<void> {
    if (!this.running) return;

    const now = this.options.now();
    const due = this.lines().filter((line) => (this.nextProbeAt.get(line.id) ?? 0) <= now);
    await Promise.all(due.map((line) => this.probe(line)));
    this.health.reconcileDegraded();

    if (this.options.bandwidthPath !== undefined) await this.maybeMeasureThroughput();

    if (!this.running) return;
    // One timer for all lines rather than one per line: the schedule is already carried by
    // nextProbeAt, and a single wakeup keeps a phone's radio asleep between rounds.
    this.timer = setTimeout(() => void this.tick(), this.options.intervalMs);
  }

  private async probe(line: Line): Promise<void> {
    const started = this.options.now();
    try {
      const response = await this.withTimeout(
        this.fetchImpl(this.resolve(this.options.probePath, line), {
          method: "GET",
          cache: "no-store",
          ...(line.url === "" ? {} : { credentials: "include" as const }),
        }),
      );
      if (!response.ok) throw new Error(`probe answered ${response.status}`);
      this.health.recordSuccess(line.id, this.options.now() - started, this.options.now());
      this.nextProbeAt.set(line.id, this.options.now() + this.options.intervalMs);
    } catch (error) {
      this.health.recordFailure(line.id, error, this.options.now());
      const failures = this.health.get(line.id).consecutiveFailures;
      this.nextProbeAt.set(line.id, this.options.now() + this.backoff(failures));
    }
  }

  /**
   * Measure throughput on one line, if the application has been quiet long enough.
   *
   * One line per round, never all of them: measuring several at once would have them compete for
   * the same pipe and each report a fraction of the truth.
   */
  private async maybeMeasureThroughput(): Promise<void> {
    const now = this.options.now();
    if (now - this.lastTrafficAt < this.options.idleBeforeBandwidthMs) return;

    const candidate = this.lines().find(
      (line) =>
        this.health.get(line.id).state !== "down" &&
        (this.nextBandwidthAt.get(line.id) ?? 0) <= now,
    );
    if (!candidate) return;

    // Claim the slot before measuring, so a failure cannot make it retry on the very next tick.
    this.nextBandwidthAt.set(candidate.id, now + this.options.bandwidthIntervalMs);

    const started = this.options.now();
    try {
      const response = await this.fetchImpl(
        this.resolve(this.options.bandwidthPath!, candidate),
        // no-store, or a second measurement would read from the disk cache and report the speed of
        // the disk.
        { method: "GET", cache: "no-store" },
      );
      const bytes = (await response.arrayBuffer()).byteLength;
      const seconds = (this.options.now() - started) / 1000;
      if (seconds > 0 && bytes > 0) this.health.recordThroughput(candidate.id, bytes / seconds);
    } catch (error) {
      this.health.recordFailure(candidate.id, error, this.options.now());
    }
  }

  /** Exponential, capped: a dead line costs one request every few minutes, not every few seconds. */
  private backoff(consecutiveFailures: number): number {
    const grown = this.options.intervalMs * 2 ** Math.min(consecutiveFailures, 10);
    return Math.min(grown, this.options.maxBackoffMs);
  }

  private withTimeout(promise: Promise<Response>): Promise<Response> {
    return new Promise((resolve, reject) => {
      const timer = setTimeout(
        () => reject(new Error(`probe timed out after ${this.options.timeoutMs}ms`)),
        this.options.timeoutMs,
      );
      promise.then(resolve, reject).finally(() => clearTimeout(timer));
    });
  }
}

File: dart/lib/src/prober.dart
/*
 *  Description: The measurement loop for Dart, ported from ts/src/prober.ts.
 *
 *               Two rules shape everything here, and they are the TypeScript ones because a client
 *               that measured its lines differently would rank them differently, and two clients
 *               disagreeing about which line is fastest is worse than neither measuring at all.
 *
 *               A probe must not distort what it measures. Latency probes are tiny and spaced out;
 *               the throughput probe downloads real bytes, so it only runs when the application has
 *               been quiet for a while — measuring bandwidth during a burst of user traffic would
 *               compete with that traffic and report a number caused by the measurement itself.
 *
 *               A failing line must not be probed at the same rate as a healthy one. A line that is
 *               down gets exponential backoff, so a dead route costs a request every few minutes
 *               rather than every few seconds.
 *
 *               Why this had to exist: the Dart port shipped with a one-shot `probe()` and nothing
 *               to call it. Real traffic only ever touches the line it was routed over, so every
 *               OTHER line stayed "never measured" forever — and a line nobody has measured cannot
 *               be ranked, which quietly turns multi-line routing back into single-line routing.
 *
 *  Author(s):
 *      Nictheboy Li    <nictheboy@outlook.com>
 */

import 'dart:async';
import 'dart:math' as math;

import 'health.dart';
import 'registry.dart';

/// What one probe request came back with.
///
/// Deliberately not a `Response`: this package does not choose an HTTP client, and the shape of the
/// answer it needs is this small.
class ProbeOutcome {
  const ProbeOutcome({required this.ok, this.bytes = 0});

  final bool ok;

  /// How many bytes arrived. Only the throughput probe cares.
  final int bytes;
}

/// Sends one probe. The consumer supplies this, because it owns the client and the credentials.
typedef SendProbe = Future<ProbeOutcome> Function(String url);

class ProberOptions {
  const ProberOptions({
    this.probePath = '/mt/probe',
    this.interval = const Duration(seconds: 15),
    this.timeout = const Duration(seconds: 5),
    this.maxBackoff = const Duration(minutes: 5),
    this.bandwidthPath,
    this.bandwidthInterval = const Duration(minutes: 10),
    this.idleBeforeBandwidth = const Duration(seconds: 30),
  });

  /// Path of the unauthenticated liveness endpoint.
  final String probePath;

  /// Gap between latency probes of a healthy line.
  final Duration interval;

  /// A probe that has not answered within this is a failure.
  final Duration timeout;

  /// Backoff cap for a line that keeps failing.
  final Duration maxBackoff;

  /// Path of a download used to measure throughput. Null disables throughput probing entirely.
  final String? bandwidthPath;

  /// Minimum gap between throughput measurements of the same line.
  final Duration bandwidthInterval;

  /// How long the application must have been quiet before a throughput probe may run.
  final Duration idleBeforeBandwidth;
}

/// Drives the health table.
///
/// Deliberately not started by the constructor: probing is a background activity with a real cost,
/// and a library that begins making network requests the moment it is instantiated is a library
/// that surprises people. The caller says when.
class Prober {
  Prober({
    required List<Line> Function() lines,
    required HealthTable health,
    required SendProbe send,
    required String Function(String path, Line line) resolve,
    this.options = const ProberOptions(),
    DateTime Function()? now,
  })  : _lines = lines,
        _health = health,
        _send = send,
        _resolve = resolve,
        _now = now ?? DateTime.now;

  final List<Line> Function() _lines;
  final HealthTable _health;
  final SendProbe _send;
  final String Function(String path, Line line) _resolve;
  final ProberOptions options;
  final DateTime Function() _now;

  /// Next allowed probe time per line, which is how backoff is expressed.
  final Map<String, DateTime> _nextProbeAt = {};
  final Map<String, DateTime> _nextBandwidthAt = {};

  /// When the application last issued a real request, so quiet can be told from busy.
  DateTime? _lastTrafficAt;

  Timer? _timer;
  bool _running = false;

  /// Called by the line manager on every real request, so the prober can tell quiet from busy.
  void noteTraffic() => _lastTrafficAt = _now();

  void start() {
    if (_running) return;
    _running = true;
    unawaited(_tick());
  }

  void stop() {
    _running = false;
    _timer?.cancel();
    _timer = null;
  }

  /// Probe everything now and wait for it. The panel's refresh button, and every test's lever.
  Future<void> probeAll() async {
    await Future.wait(_lines().map(_probe));
    _health.reconcileDegraded();
  }

  Future<void> _tick() async {
    if (!_running) return;

    final now = _now();
    final due = _lines()
        .where((line) => !(_nextProbeAt[line.id]?.isAfter(now) ?? false))
        .toList();
    await Future.wait(due.map(_probe));
    _health.reconcileDegraded();

    if (options.bandwidthPath != null) await _maybeMeasureThroughput();

    if (!_running) return;
    // One timer for all lines rather than one per line: the schedule is already carried by
    // _nextProbeAt, and a single wakeup keeps a phone's radio asleep between rounds.
    _timer = Timer(options.interval, () => unawaited(_tick()));
  }

  Future<void> _probe(Line line) async {
    final started = _now();
    try {
      final outcome = await _send(
        _resolve(options.probePath, line),
      ).timeout(options.timeout);
      if (!outcome.ok) throw StateError('probe for ${line.id} was refused');
      _health.recordSuccess(line.id, _now().difference(started), _now());
      _nextProbeAt[line.id] = _now().add(options.interval);
    } catch (error) {
      _health.recordFailure(line.id, error, _now());
      _nextProbeAt[line.id] = _now().add(
        _backoff(_health[line.id].consecutiveFailures),
      );
    }
  }

  /// Measure throughput on one line, if the application has been quiet long enough.
  ///
  /// One line per round, never all of them: measuring several at once would have them compete for
  /// the same pipe and each report a fraction of the truth.
  Future<void> _maybeMeasureThroughput() async {
    final now = _now();
    final quietSince = _lastTrafficAt;
    if (quietSince != null &&
        now.difference(quietSince) < options.idleBeforeBandwidth) {
      return;
    }

    Line? candidate;
    for (final line in _lines()) {
      final ready = !(_nextBandwidthAt[line.id]?.isAfter(now) ?? false);
      if (_health[line.id].state != LineState.down && ready) {
        candidate = line;
        break;
      }
    }
    if (candidate == null) return;

    // Claim the slot before measuring, so a failure cannot make it retry on the very next tick.
    _nextBandwidthAt[candidate.id] = now.add(options.bandwidthInterval);

    final started = _now();
    try {
      final outcome = await _send(_resolve(options.bandwidthPath!, candidate));
      final seconds = _now().difference(started).inMicroseconds / 1000000;
      if (seconds > 0 && outcome.bytes > 0) {
        _health.recordThroughput(candidate.id, outcome.bytes / seconds);
      }
    } catch (error) {
      _health.recordFailure(candidate.id, error, _now());
    }
  }

  /// Exponential, capped: a dead line costs one request every few minutes, not every few seconds.
  Duration _backoff(int consecutiveFailures) {
    final grown = options.interval *
        math.pow(2, math.min(consecutiveFailures, 10)).toDouble();
    return grown > options.maxBackoff ? options.maxBackoff : grown;
  }
}
